src/stream/src/common/table/state_table.rs

```diff
@@ -938,6 +937,34 @@ where
             .map(get_second))
     }
 
+    async fn iter_all_pk_and_val_with_pk_range(
```

Comment:
I remembered the storage table has a logic to transform a `Exclude` begin key into a `Include` begin key, for the reason that the storage doesn't support it yet. Is this logic also necessary for the state table? BTW, why storage doesn't support `Exclude` begin key ? cc @BugenZhao cc @wenym1 
https://github.com/risingwavelabs/risingwave/blob/1392d4b2d6e05a37865d399a96ca69ed1d602627/src/storage/src/table/batch_table/storage_table.rs#L548-L563

Reply:
I find the state table contains this logic as well. So far so good.

Reply:
Internally state table goes through the following steps:
1. iter_all_pk_and_val_with_pk_range
2. ...
3. iter_with_pk_range_inner.
4. Compute range: `prefix_range_to_memcomparable`.
5. Further, that calls `to_memcomparable` on upper and lower bounds.
6. The definition of `to_memcomparable`:
```rs

fn to_memcomparable<R: Row>(
    pk_serde: &OrderedRowSerde,
    bound: &Bound<R>,
    is_upper: bool,
) -> Bound<Bytes> {
    let serialize_pk_prefix = |pk_prefix: &R| {
        let prefix_serializer = pk_serde.prefix(pk_prefix.len());
        serialize_pk(pk_prefix, &prefix_serializer)
    };
    match bound {
        Unbounded => Unbounded,
        Included(r) => {
            let serialized = serialize_pk_prefix(r);
            if is_upper {
                end_bound_of_prefix(&serialized)
            } else {
                Included(serialized)
            }
        }
        Excluded(r) => {
            let serialized = serialize_pk_prefix(r);
            if !is_upper {
                // if lower
                start_bound_of_excluded_prefix(&serialized)
            } else {
                Excluded(serialized)
            }
        }
    }
}
```

For `Included` the logic looks like it's the same, with the exception that it handles `pk_prefix`. For our case `pk_prefix` is empty, so we can ignore it:
```
        let iter = upstream_table
            .batch_iter_with_pk_bounds(
                HummockReadEpoch::NoWait(epoch),
                row::empty(), //  <------------- empty
                range_bounds,
                ordered,
                PrefetchOptions::new_for_exhaust_iter(),
            )
            .await?
            .map(get_second);
```

For `Excluded` the logic is also similar, one calls `next_key`, the other calls `start_bound_of_excluded_prefix`. But both definitions are similar, it's just that `next_key` does not construct the bound inside. 
```
pub fn next_key(key: &[u8]) -> Vec<u8> {
    if let Some((s, e)) = next_key_no_alloc(key) {
        let mut res = Vec::with_capacity(s.len() + 1);
        res.extend_from_slice(s);
        res.push(e);
        res
    } else {
        Vec::new()
    }
}

/// Get the start bound of the given `prefix` when it is excluded from the range.
pub fn start_bound_of_excluded_prefix(prefix: &[u8]) -> Bound<Bytes> {
    if let Some((s, e)) = next_key_no_alloc(prefix) {
        let mut buf = BytesMut::with_capacity(s.len() + 1);
        buf.extend_from_slice(s);
        buf.put_u8(e);
        Included(buf.freeze())
    } else {
        panic!("the prefix is the maximum value")
    }
}

```

So perhaps it can be unified to some extent, but not totally, since there is extra logic in batch side to handle a pk_prefix on top of pk_range.

Reply:
> I find the state table contains this logic as well. So far so good.

Just saw your comment 😆 , but I think we can refactor it still in separate PR.